File: 01-data-structures/stack/stack.cpp
#include <iostream>
#include <stack>
#include <list>
using namespace std;
int main()
{
  //构建 stack 容器适配器
  list<int> values = {1, 2, 3, 4, 5, 6};
  stack<int, list<int>> my_stack(values);
  //查看 my_stack 存储元素的个数
  cout << "size of my_stack: " << my_stack.size() << endl;
  cout << "my_stack.pop(): " << endl;
  // 弹出1个
  my_stack.pop();
  cout << "my_stack.top():" << my_stack.top() << endl;
  // 压栈1个
  my_stack.push(4);
  //将 my_stack 中存储的元素依次弹栈，直到其为空
  while (!my_stack.empty())
  {
    cout << "item:" << my_stack.top() << endl;
    //将栈顶元素弹栈
    my_stack.pop();
  }
  return 0;
}

/**
jarrys-MacBook-Pro:stack jarry$ g++ -std=c++11 -o stack stack.cpp
jarrys-MacBook-Pro:stack jarry$ ./stack 
size of my_stack: 3
my_stack.pop(): 
2
item:4
item:2
item:1
 */

File: 01-data-structures/stack/stack.js
class Stack {
  constructor(size) {
    this.top = 0;
    this.size = size;
    this.data = new Array(size);
  }

  // 打印栈中的所有元素
  printStack() {
    if (this.isEmpty()) {
      console.log("栈为空");
      return;
    }

    for (let i = 0; i < this.top; i++) {
      process.stdout.write(this.data[i] + " ");
    }
    console.log();
  }

  // 判断栈是否已满
  isFull() {
    return this.top === this.size;
  }

  // 判断栈是否为空
  isEmpty() {
    return this.top === 0;
  }

  // 入栈操作
  push(value) {
    if (this.isFull()) {
      return false;
    }

    this.data[this.top] = value;
    this.top++;
    return true;
  }

  // 出栈操作
  pop() {
    if (this.isEmpty()) {
      return 0;
    }

    this.top--;
    return this.data[this.top];
  }

  // 清空栈
  makeEmpty() {
    this.top = 0;
  }

  // 获取栈中元素的个数
  len() {
    return this.top;
  }
}

// 测试栈功能
function testStack() {
  let s = new Stack(5);
  console.log("执行入栈操作");
  console.log("入栈 1:", s.push(1));
  s.printStack();
  console.log("入栈 2:", s.push(2));
  s.printStack();

  console.log("执行出栈操作");
  console.log("出栈:", s.pop());
  s.printStack();
  console.log("出栈:", s.pop());
  s.printStack();

  console.log("清空栈");
  s.makeEmpty();
  s.printStack();

  console.log("测试结束");
}

testStack();

/*
jarry@MacBook-Pro stack % node stack.js
执行入栈操作
入栈 1: true
1 
入栈 2: true
1 2 
执行出栈操作
出栈: 2
1 
出栈: 1
栈为空
清空栈
栈为空
测试结束
*/